text/0732-cdk-cli-telemetry.md

```diff
@@ -0,0 +1,272 @@
+
+When something goes wrong in the CDK CLI, we will report the error to the telemetry service.
+The error schema looks like this:
+
+```ts
+export type ErrorDetails = {
+  name: string;
+  message?: string; // anonymized error message
+  trace?: string; // anonymized stack trace
+  logs?: string; // anonymized stack logs
+}
+```
+
+The error name is an enum — currently either `ToolkitError`, `AuthorizationError`, `AssemblyError`, etc.
+
+The error message, trace, and logs could contain customer input.
+We have compiled the following list of possible customer inputs found in these strings and what we will do to redact those inputs:
+
+#### ARNs
+
+An ARN is a string that is prefixed by `arn:`. `arn:aws:s3:::my-bucket` will turn into `<redacted-arn>`.
+
+#### AccountIDs
+
+Anything that is a 12 digit number (that wasn’t already redacted within an ARN) will be redacted. `123456789012` will turn into `<redacted-account-id>`.
+
+#### File Paths
+
+Part of the file path (before node_modules) is customer content that we do not need to store.
+We will use a regex that includes the following to find and replace file path prefixes:
+
+```
+(file:/+)? -> matches optional file url prefix
+homedir()/process.cwd() -> users home directory or current working directory, replacing \ with /
+[\\w.\\-_@\\\\/]+ -> matches nested directories and file name
+```
+
+For example, a file path like `~/MyUser/my/path/to/node_modules/cdk-assets/lib/asset-manifest.js:20:19` will be truncated to `node_modules/cdk-assets/lib/asset-manifest.js:20:19`.
+
+#### Stack Names
+
+Stack Names are arbitrary customer content that we will redact.
+The CDK CLI will know the names of all the Stack Names when the Cloud Assembly is synthesized.
+That means that, for every command that undergoes synthesis, we will have all Stack IDs available.
+The CLI will collect these names and then find and redact them in the error message, stack trace, and logs.
+
+The commands that undergo synthesis are `cdk list`, `cdk diff`, `cdk synth`, and `cdk deploy`.
+**For all other commands, since we do not know what could be the Stack Name, we will not collect error messages, traces, or logs.**
+
+For example, the stack name might show up in the following string: `'cdk.out/MyStackName.assets.json'`.
+Once we have synthesized, we will know that MyStackName is a potential stack name and redact.
+The string will become `cdk.out/<redacted-stack-name>.assets.json`.
+
+#### Asset Display Names
+
+Asset Display Names are arbitrary customer content that we will redact.
+The CDK CLI will know the names of the Asset Display Names in the `.assets.json`
+file after the Cloud Assembly is synthesized.
+Similar to Stack Names, we will collect and redact these Asset Names and
+**will not collect error messages, traces, or logs when the cloud assembly is unavailable.**
+
+A log message like `start: Building TelemetryFunction/Code` becomes `start: Building <redacted-asset-name>`.
```

Comment:
Still don't understand why we wouldn't replace with `****`.

For something more elaborate: it would be nice if we would replace redacted strings with symbolic names. `$ACCOUNT1`, `$ACCOUNT2`, `$ASSET1`, etc. That way, if there are multiple entities of the same type, we can distinguish them, and if the same entity gets referenced multiple times in the same log, we can cross-reference the occurrences.

Reply:
Good idea. For assets and stacks, I can do this since I'm getting the names from the cloud assembly. For accounts and arns, it's more of a blanket sanitation and i would have to store information on which values I've found in order to count them `$ACCOUNT1` `$ACCOUNT2`, etc. read that as more work than I want to commit to right now, and we can start with `$STACK1` and `$ASSET1` but no counts for other things